### ISSUE TITLE
Update Sampler class to maintain state (among other things)

### ISSUE THREAD

**Issue by nabriis** [CUQI-DTU/CUQIpy#34]
Currently most of our samplers are not able to maintain their current state and continue sampling. Instead they all implement their own `sample` method which initializes, runs and outputs sample results. This makes it very difficult and inefficient to implement true Gibbs sampling using our current samplers. To remidy this we would need a total rework of the base sampler class to implement something like the following:

- Allocate sample method
- warmup method (samples in a warmup/burnin phase and tunes parameters)
- sample method
- output samples method
- Printing progress

The subclassing samplers would be required to implement the following
- step (take a single MCMC step from the current position)
- tune (tune parameters according to the current state)

# Comments
jakj: It is neat to write out progress by updating the same one line. On the other hand printing a new line each time allows us to add more information so as to monitor progress over iterations, even just adding time spent pr sample is helpful to allow monitoring if this changes and how long to expect it still will take.

amal: 
- samplers diagnostics can be considered #274 for all samplers (we have some diagnostics in NUTS for now) 
- related issue #147 

# Babak suggested code structure
See code for a suggested structure for the samplers that are entirely stateful (i.e. they keep track of their state).

``` Python
import numpy as np
import matplotlib.pyplot as plt

class Sampler():
    def __init__(self, log_pi_target, x0):
        self.log_pi_target = log_pi_target
        self.dim = x0.shape[0]
        self.x_old = x0
        self.target_old = self.log_pi_target( self.x_old )

        self.samples = [ x0 ]
        self.targets = [ self.target_old ]
        self.acc = [ 1. ]

    def sample(self, Ns):
        for i in range(Ns):
            local_acc = self.step()
            self.samples.append( self.x_old )
            self.targets.append( self.target_old )
            self.acc.append( local_acc )


class Metropolis_Hastings(Sampler):
    def __init__(self,log_pi_target, x0, scale=0.1):
        super().__init__(log_pi_target, x0)
        self.scale = 0.1

        self.num_adapt = 0
        self.star_acc = 0.234    # target acceptance rate RW

    def set_pi_target(self, log_pi_target):
        self.log_pi_target = log_pi_target

    def set_scale(self, scale):
        self.scale = scale

    def step(self):
        x = self.x_old + self.scale*np.random.standard_normal(self.dim)

        target = self.log_pi_target( x )
        ratio = target - self.target_old
        alpha = min(0., ratio)
        uu = np.log( np.random.rand() )
        if (uu <= alpha):
            acc = 1
            self.x_old = x
            self.target_old = target
        else:
            acc = 0

        return acc

    def tune(self, history_length=1):
        av_acc = np.mean( self.acc[-history_length:] )
        zeta = 1/np.sqrt(self.num_adapt+1)
        self.scale = np.exp(np.log(self.scale) + zeta*(av_acc-self.star_acc))
        print(self.scale)



if __name__ == '__main__':
    log_pi = lambda p: -0.5*p**2
    x0 = np.array( [0.] )
    sampler = Metropolis_Hastings(log_pi,x0)
    sampler.scale

    sampler.sample(500)
    sampler.tune(history_length=50)
    sampler.sample(500)
    sampler.tune(history_length=50)

    samps = np.array( sampler.samples ).flatten()
    plt.hist(samps, bins=100)
    plt.show()
```

**Comment by nabriis** [comment]
We agreed to split this task into smaller subtasks
- [ ] Add new baseclass sampler + MH + pCN.
- [ ] Linear_RTO + UGLA sampler.
- [ ] CWMH
- [ ] ULA, MALA
- [ ] NUTS
- [ ] Gibbs

**Comment by nabriis** [comment]
Finally closed by #425
